core/interop/Streams/ManagedLib/StreamAsCExports.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace InteropWithStream;
+
+/// <summary>
+/// Stream data type projected as an opaque handle and interacted via C exports.
+/// </summary>
+public static unsafe partial class StreamExports
+{
+#if NETFRAMEWORK
+    [DNNE.Export]
+#else
+    [UnmanagedCallersOnly]
+#endif // !NETFRAMEWORK
+    public static IntPtr C_CreateStream()
+    {
+        MemoryStream stream = new();
+        return GCHandle.ToIntPtr(GCHandle.Alloc(stream));
+    }
+
+#if NETFRAMEWORK
+    [DNNE.Export]
+#else
+    [UnmanagedCallersOnly]
+#endif // !NETFRAMEWORK
+    public static void C_DeleteStream(IntPtr streamMaybe)
+    {
+        if (streamMaybe != IntPtr.Zero)
+        {
+            GCHandle.FromIntPtr(streamMaybe).Free();
+        }
+    }
+
+#if NETFRAMEWORK
+    [DNNE.Export]
+#else
+    [UnmanagedCallersOnly]
+#endif // !NETFRAMEWORK
+    public static int C_PrintStream(IntPtr streamMaybe)
+    {
+        try
+        {
+            var stream = (MemoryStream)GCHandle.FromIntPtr(streamMaybe).Target!;
+            var data = stream.ToArray();
+            const int row = 4;
+            for (int i = 0; i < data.Length; ++i)
+            {
+                char delim = (i % row == row - 1) ? '\n' : ' ';
+                Console.Write($"{data[i]}{delim}");
+            }
+            Console.WriteLine();
+        }
+        catch (Exception e)
+        {
+            return e.HResult;
+        }
+        return 0;
+    }
+
+    //
+    // APIs for interacting with Streams via C export
+    //
+
+#if NETFRAMEWORK
+    [DNNE.Export]
+#else
+    [UnmanagedCallersOnly]
+#endif // !NETFRAMEWORK
+    public static int C_Stream_Read(IntPtr streamMaybe, nint length, byte* dataRaw)
+    {
+        Debug.Assert(length < int.MaxValue);
+        try
+        {
+            var stream = (Stream)GCHandle.FromIntPtr(streamMaybe).Target!;
+            Span<byte> data = new(dataRaw, (int)length);
+#if NETFRAMEWORK
+            var buffer = new byte[data.Length];
+            stream.Read(buffer, 0, buffer.Length);
+            buffer.CopyTo(data);
+#else
+            stream.Read(data);
```

Comment:
These Read calls are not guaranteed to read `length` bytes.  And the shape of this C_Stream_Read method means the caller doesn't know how many bytes were actually read.